File: al-leetcode/dp_range-sum-query-2d-immutable.java
// https://www.youtube.com/watch?v=zvqJMGWjuX0
// https://leetcode.com/problems/range-sum-query-2d-immutable/submissions/

// Given a 2D matrix matrix, handle multiple queries of the following type:

// Calculate the sum of the elements of matrix inside the rectangle defined by its upper left corner (row1, col1) and lower right corner (row2, col2).
// Implement the NumMatrix class:

// NumMatrix(int[][] matrix) Initializes the object with the integer matrix matrix.
// int sumRegion(int row1, int col1, int row2, int col2) Returns the sum of the elements of matrix inside the rectangle defined by its upper left corner (row1, col1) and lower right corner (row2, col2).
class NumMatrix {

    private int[][] sum;
    private int[][] mat;
    private int m;
    private int n;
    private int[][] res;
    

    public NumMatrix(int[][] matrix) {
        // NumMatrix(int[][] matrix) Initializes the object with the integer matrix matrix.
        
        int m = matrix.length;
        int n = matrix[0].length;
        this.m = m;
        this.n = n;
        this.mat = matrix;
        this.sum = new int[m + 1][n + 1];
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                sum[i][j] = sum[i - 1][j] + sum[i][j - 1] - sum[i - 1][j - 1] + matrix[i - 1][j - 1];
            }
        }
    }
    
    public int sumRegion(int row1, int col1, int row2, int col2) {
        // int sumRegion(int row1, int col1, int row2, int col2) Returns the sum of the elements of matrix inside the rectangle defined by its upper left corner (row1, col1) and lower right corner (row2, col2).
        
        return sum[row2 + 1][col2 + 1] - sum[row2 + 1][col1] - sum[row1][col2 + 1] + sum[row1][col1];        
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.sumRegion(row1,col1,row2,col2);
 */

File: al-leetcode/uncategorized/dp_range-sum-query-2d-immutable.js
/**
 * @param {number[][]} matrix
 */
var NumMatrix = function(matrix) {

    this.matrix = matrix;
    this.sum = [];
    for (let i = 0; i < matrix.length; i++) {
        let row = [];
        for (let j = 0; j < matrix[i].length; j++) {
            if (i === 0) {
                row.push(matrix[i][j]);
            } else {
                row.push(row[j - 1] + matrix[i][j]);
            }
        }
        this.sum.push(row);
    }
};

/** 
 * @param {number} row1 
 * @param {number} col1 
 * @param {number} row2 
 * @param {number} col2
 * @return {number}
 */
//  int sumRegion(int row1, int col1, int row2, int col2) Returns the sum of the elements of matrix inside the rectangle defined by its upper left corner (row1, col1) and lower right corner (row2, col2).

NumMatrix.prototype.sumRegion = function(row1, col1, row2, col2) {
    let sum = 0;
    // O(1)
    for (let i = row1; i <= row2; i++) {
        sum += this.sum[i][col2] - (col1 > 0 ? this.sum[i][col1 - 1] : 0);
    }
    return sum;
};

/** 
 * Your NumMatrix object will be instantiated and called as such:
 * var obj = new NumMatrix(matrix)
 * var param_1 = obj.sumRegion(row1,col1,row2,col2)
 */